src/media/hwschematic.ts

```diff
@@ -146,43 +164,175 @@ const findWidgetFormState = new FindWidgetFormData(
+	// }
+
 	hwSchematic.bindData(json).then(
 		function () { return; },
 		(e: any) => {
 			hwSchematic.setErrorText(e);
 			throw e;
 		});
+
+		// if (json.hwMeta.highlightGroups !== undefined && json.hwMeta.highlightGroups !== null) {
+		// 	const nodes: any[] = getElementsToAdd(hwSchematic.layouter.graph, (item: any) => true);
+		// 	const currentlySelectedIds = new Set();
+		// 	for () {
+
+		// 	}
+		
+		// }
+
+		if ((eventType == "init" || eventType == "") && json.hwMeta.highlightGroups !== undefined && json.hwMeta.highlightGroups !== null) {
+			currentlySelected.clear();
+			const nodes: any[] = getElementsToAdd(hwSchematic.layouter.graph, (item: any) => true);
+			const currentlySelectedIds = getCurrentlySelectedIds(json.hwMeta.highlightGroups);
+
+			for (const node of nodes) {
+				if (currentlySelectedIds.has(node.id)) {
+					currentlySelected.add(node);
+				}
+			}
+			applyHighlight(false);
+			addHighlightGroups(nodes, json.hwMeta.highlightGroups);
+		}
+
+	
+
+
+
+}
+
+/*
+ Checks two object for equality (recursively)
+*/
+function isDeepEqual(object1: any, object2: any) {
+	const objKeys1 = Object.keys(object1);
+	const objKeys2 = Object.keys(object2);
+
+	if (objKeys1.length !== objKeys2.length) return false;
+
+	for (const key of objKeys1) {
+		const value1 = object1[key];
+		const value2 = object2[key];
+
+		const isObjects = isObject(value1) && isObject(value2);
+
+		if ((isObjects && !isDeepEqual(value1, value2)) ||
+			(!isObjects && value1 !== value2)
+		) {
+			return false;
+		}
+	}
+	return true;
+}
+
+function isObject(object: any) {
+	return object != null && typeof object === "object";
 }
 
 // Handle messages sent from the extension to the webview
 window.addEventListener('message', (event) => {
-	const message = event.data; // The json data that the extension sent
-	switch (message.type) {
+	const {type, body} = event.data;
+	switch (type) {
+		case 'init':
 		case 'update': {
-			const text = message.text;
+			const textBytes = body.json;
+			const decoder = new TextDecoder();
+			const text = decoder.decode(textBytes);
+			let json: any = {};
+			try {
+				json = JSON.parse(text);
+			} catch {
+				if (hwSchematic)
+					hwSchematic.setErrorText('Error: Document is not valid json');
+			}
+
 			const state = vscode.getState();
-			if (!state || state.text != text) {
+			if (!state || isDeepEqual(state.json, json)) {
 				// Update our webview's content
-				updateContent(text);
+				// copy json for updateContent because it modifies it
+				updateContent(JSON.parse(JSON.stringify(json)), type);
+				//json.hwMeta.highlightGroups = findWidgetFormState.getHighlightGroups();
+
 			}
 			// Then persist state information.
 			// This state is returned in the call to `vscode.getState` below when a webview is reloaded.
-			vscode.setState({ text });
+			vscode.setState({ json: json });
+			return;
+		}
+
+		case 'getFileData':
+		{
+			// Get the image data for the canvas and post it back to the extension.
```

Comment:
Update comments from original PawDraw example and add reference to in into file header or class doc